src/workerd/io/io-context.c++

```diff
@@ -1082,7 +1082,47 @@ void IoContext::runImpl(Runnable& runnable, bool takePendingEvent,
       js.terminateExecution();
     }
 
-    js.runMicrotasks();
+    // Running the microtask queue can itself trigger a pending exception in the isolate.
+    v8::TryCatch tryCatch(workerLock.getIsolate());
+
+    try {
+      js.runMicrotasks();
+    } catch (jsg::JsExceptionThrown&) {
+      // We'll let the tryCatch.HasCaught below handle it.
+      if (!tryCatch.HasCaught()) {
+        LOG_NOSENTRY(WARNING, "JsExceptionThrown thrown without a pending exception?");
+      }
+    }
+
+    if (tryCatch.HasCaught()) {
+      if (tryCatch.HasTerminated() || tryCatch.Message().IsEmpty()) {
```

Comment:
I think you could assert `tryCatch.HasTerminated()` here -- a regular exception cannot be thrown from microtasks.

I don't think all the code to `logUncaughtException` below is needed. We should already be reporting the problem by other means.

I think this whole `if (tryCatch.HasCaught())` block could probably be replaced with simply `tryCatch.Reset()`?

Reply:
Fair enough